server/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -147,6 +148,20 @@ protected Engine(EngineConfig engineConfig) {
         this.eventListener = engineConfig.getEventListener();
     }
 
+    /**
+     * Reads an {@code IndexVersion} from an {@code es_version} metadata string
+     */
+    public static IndexVersion readIndexVersion(String esVersion) {
+        if (esVersion.contains(".")) {
+            // backwards-compatible Version-style
+            org.elasticsearch.Version v = org.elasticsearch.Version.fromString(esVersion);
+            assert v.before(org.elasticsearch.Version.V_8_11_0);
+            return IndexVersion.fromId(v.id);
+        } else {
+            return IndexVersion.fromId(Integer.parseInt(esVersion));
```

Comment:
Just checking: in case of problems of any kind in the format this will throw NumberFormatException. Are we OK with that?

Reply:
It'll either be version-style or indexversion-style. If its indexversion-style, it means it was created or updated by a node version that uses IndexVersion here. And indices don't go backwards in version, so once changed to IndexVersion it will always be IndexVersion. An older node version shouldn't even attempt to read an index shard created by a newer node version.

Reply:
Thanks for clarifying!